### PR TITLE
Store DateTimes with ZoneIds

### PR DESCRIPTION
**Pull request by sherfert**
This implements the storage of DateTimes with ZoneIds for both scalar
and array values. It stores ZoneIds as shorts, and uses a hardcoded
table to map between String and short.

changelog: Any temporal value can be stored in a property. This includes Date, Time, DateTime, LocalTime, LocalDateTime, and Duration.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java
Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java]: These are the wrong collections to use here.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestPropertyTypes.java
Comment (thobe) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestPropertyTypes.java]: This is insufficient. Since it is possible for end users to update the Time Zone Database in their JRE, we really ought to update the time zone mapping every time there is a new release of the IANA Time Zone Database.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestPropertyTypes.java
Comment (thobe) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestPropertyTypes.java]: This assertion is insufficient. We should assert on the actual table of tzids, since even official JRE releases will update the Time Zone Database on patch releases.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java
Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java]: It really irks me to have this defined in code. I'd much prefer to have this defined in a resource file that is read when this class loads.

Keep in mind that we also need to define mappings for tzids that have been removed from the tz database.

**Review by thobe — CHANGES_REQUESTED** [review]
There are a few issues with this. But most importantly I'm missing a strategy for how the mapping of tzids is supposed to be kept up to date.

I have written up a short document on the things to keep in mind whit this sort of mapping: https://docs.google.com/document/d/1FS0V7B3oj09xOb23KXNXHufkjBjZxedcIeq2Qtv4pbs/edit?usp=sharing

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java]: Probably better with something like a [PrimitiveIntObjectMap](https://github.com/neo4j/neo4j/blob/3.4/community/primitive-collections/src/main/java/org/neo4j/collection/primitive/Primitive.java#L181)

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TemporalType.java
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/TemporalType.java]: Has this `3` any relation to the one in `calculateNumberOfBlocksUsedForTemporal`? If so, or perhaps even if not, then introducing constants is a good idea and really aids readers of this code.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TemporalType.java
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/TemporalType.java]: A little outline of what's in these 3 blocks as a comment would be dandy

**Review by tinwelint — CHANGES_REQUESTED** [review]

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java
Comment (sherfert) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java]: Which ones would be the right ones?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TemporalType.java
Comment (sherfert) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/store/TemporalType.java]: I will put that into the comment in PropertyStore, where all store layouts are summarized. Thanks for the reminder!

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java
Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java]: Not the only reason why we could get an exception here. I'd reword this to just `"Failed to read time zone id file."`.

File: community/kernel/src/main/resources/TZIDS
Comment (thobe) [inline community/kernel/src/main/resources/TZIDS]: I think there should be a newline at the end. Even if that means that the reader code needs to know how to skip blank lines.

File: community/kernel/src/main/resources/TZIDS
Comment (thobe) [inline community/kernel/src/main/resources/TZIDS:587]: _This is not a suggestion, it is a question._

For the initial chunk - the identifiers before `tzdata2014b` - how about we order them in a way that makes this file easier for humans?

The argument for keeping it unordered would be that it would not tempt maintainers to try to insert new entries "in order", in stead of at the end as they should.
The argument for having an order would be that it makes debugging easier.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/store/TimeZoneMappingTest.java
Comment (thobe) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/store/TimeZoneMappingTest.java]: There is an alternative URL for this under the `data` domain that we might want to use instead: `https://data.iana.org/time-zones/tzdata-latest.tar.gz`.

Comment (thobe) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/store/TimeZoneMappingTest.java]: We should give this variable a better name than `foo`.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/store/TimeZoneMappingTest.java
Comment (thobe) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/store/TimeZoneMappingTest.java]: I'd like to factor this code into a some helper method:
```
void readTZDBFile( String release, String file, Predicate<String> consumer ) throws IOException
{
    String uri = release == null
            ? "https://data.iana.org/time-zones/tzdata-latest.tar.gz"
            : "https://data.iana.org/time-zones/releases/tzdata" + release + ".tar.gz";
    try ( TarArchiveInputStream tar = new TarArchiveInputStream( new GZIPInputStream( new URL( uri ).openStream() ) ) )
    {
        for ( ArchiveEntry entry; (entry = tar.getNextEntry()) != null; )
        {
            if ( file.equalsIgnoreCase( entry.getName() ) )
            {
                try ( BufferedReader reader = new BufferedReader( new InputStreamReader( tar ) ) )
                {
                    for ( String line; (line = reader.readLine()) != null; )
                    {
                        if ( !consumer.test( line ) )
                        {
                            break;
                        }
                    }
                }
                break;
            }
        }
    }
}
```

Then our two usages of this pattern would become:
```
readTZDBFile( null, "NEWS", line ->
{
    if ( line.startsWith( "Release " ) )
    {
        String substring = line.substring( line.indexOf( ' ' ) + 1 );
        String release = substring.substring( 0, substring.indexOf( ' ' ) );
        if ( TimeZoneMapping.LATEST_SUPPORTED_IANA_VERSION.equals( release ) )
        {
            return false; // stop reading
        }
        versionsToUpgrade.add( release );
    }
    return true; // continue reading
} );
```
and:
```
readTZDBFile( release, "zone1970.tab", line ->
{
    if ( !line.startsWith( "#" ) )
    {
        ianaSupportedTzs.add( line.split( "\\t" )[2] );
    }
    return true; // read all lines
} );
```

Which is much easier on the eyes (even if it does need the comments for explaining how the return value is used for control flow).

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java
Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/TimeZoneMapping.java:43]: I'd like to change `TIME_ZONE_SHORT_TO_STRING` to just be an array, and I think we can introduce a lower overhead data structure for `TIME_ZONE_STRING_TO_SHORT`, but those are future improvements that can be made without breaking any functionality.

**Review by thobe — APPROVED** [review]
I think this is good enough. There are still some questions I'd like to discuss before this is merged, but overall this is good enough.

There are also some things that can be further improved, but that doesn't need to stop this from going in.

File: community/kernel/src/main/resources/TZIDS
Comment (sherfert) [reply community/kernel/src/main/resources/TZIDS]: It does already know how to do that

**Review by tinwelint — APPROVED** [review]

File: community/kernel/src/main/resources/TZIDS
Comment (sherfert) [reply community/kernel/src/main/resources/TZIDS:587]: I have no strong preference for one or the other.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/store/TimeZoneMappingTest.java
Comment (sherfert) [reply community/kernel/src/test/java/org/neo4j/kernel/impl/store/TimeZoneMappingTest.java]: Agreed

**Pull request merged**
